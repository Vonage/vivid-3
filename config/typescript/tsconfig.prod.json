{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "../../dist",
    "baseUrl": "../../src"
  },
  "exclude": [
    "../../src/**/*.spec.ts",
    "../../src/**/*.test.ts"
  ]
}
