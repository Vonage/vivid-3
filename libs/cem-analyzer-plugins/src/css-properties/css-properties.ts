import * as fs from 'fs';
import * as sass from 'sass';
import * as path from 'path';
import type { Plugin } from '@custom-elements-manifest/analyzer';
import type {
	CssCustomProperty,
	CustomElementDeclaration,
} from 'custom-elements-manifest';
import { exports as resolveExports } from 'resolve.exports';

const resolveScssPath = (importPath: string): string => {
	// e.g. 'path/module' -> 'path/_module'
	const prefixed = path.join(
		path.dirname(importPath),
		`_${path.basename(importPath)}`
	);
	for (const candidate of [
		importPath,
		`${importPath}.scss`,
		`${importPath}/index.scss`,
		prefixed,
		`${prefixed}.scss`,
		`${importPath}/_index.scss`,
	]) {
		if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
			return candidate;
		}
	}
	throw new Error(`SCSS file not found for path: ${importPath}`);
};

/// Custom importer for Sass to resolve package paths
const importer: sass.Importer = {
	canonicalize(url, options) {
		if (url.startsWith('@repo')) {
			const libName = url.split('/')[1]; // e.g. 'styles'
			const pkg = JSON.parse(
				fs.readFileSync(
					path.resolve(path.join('..', libName, 'package.json')),
					'utf-8'
				)
			);
			const [resolved] = resolveExports(
				pkg,
				`./${url.split('/').slice(2).join('/')}`,
				{
					conditions: ['sass'],
				}
			)!;
			const resolvedPath = path.resolve(path.join('..', libName, resolved));
			return new URL(`file://${resolveScssPath(resolvedPath)}`);
		}

		if (options.containingUrl) {
			const resolvedPath = path.resolve(
				path.dirname(options.containingUrl.pathname),
				url
			);
			return new URL(`file://${resolveScssPath(resolvedPath)}`);
		}
		return null;
	},
	load(url: URL) {
		return {
			contents: fs.readFileSync(url, 'utf-8'),
			syntax: 'scss',
		};
	},
};

/**
 * Plugin to extract CSS custom properties from the components SCSS file.
 */
export const cssPropertiesPlugin = (): Plugin => {
	const compiler = sass.initCompiler();

	return {
		name: 'stylePlugin',
		analyzePhase({ ts, node, moduleDoc }) {
			switch (node.kind) {
				case ts.SyntaxKind.ClassDeclaration: {
					const className = (node as any).name.getText();
					const classDeclaration = moduleDoc.declarations!.find(
						(declaration) => declaration.name === className
					) as CustomElementDeclaration;

					// Try to find corresponding scss file
					const filePath = node.getSourceFile().fileName; // e.g 'libs/components/src/lib/tree-view/tree-view.ts'
					const withoutExtension = filePath.replace(/\.[^/.]+$/, '');
					const stylePath = `${withoutExtension}.scss`;
					if (!fs.existsSync(stylePath)) {
						return;
					}

					const result = compiler.compile(stylePath, {
						importers: [importer, new sass.NodePackageImporter()],
					});
					const compiledCss = result.css.toString();

					for (const line of compiledCss.split('\n')) {
						const match = line.match(
							/^.*\/\*\s*@cssprop(erty)?\s+(.+)\s*\*\/\s*$/
						);
						if (!match) {
							continue;
						}

						let definition = match[2];

						const cssProperty: Partial<CssCustomProperty> = {};

						if (definition.startsWith('[')) {
							// Definition provides a default value, e.g. '[--vvd-button-cta-primary=var(--vvd-color-canvas-text)]'
							const insideBrackets = definition.substring(
								1,
								definition.indexOf(']')
							);
							definition = definition
								.substring(definition.indexOf(']') + 1)
								.trim();
							const parts = insideBrackets.split('='); // e.g. ['--vvd-button-cta-primary', 'var(--vvd-color-canvas-text)']
							cssProperty.name = parts[0]; // e.g. '--vvd-button-cta-primary'
							cssProperty.default = parts[1]; // e.g. 'var(--vvd-color-canvas-text)'
						} else {
							// Definition just contains the name, e.g. '--vvd-button-cta-primary'
							const [varName, ...rest] = definition.split(' ');
							definition = rest.join(' ').trim();
							cssProperty.name = varName; // e.g. '--vvd-button-cta-primary'
						}

						// Now there may be a description left, e.g. '- Primary button color'
						const description = definition.replace('-', '').trim();
						if (description) {
							cssProperty.description = description;
						}

						classDeclaration.cssProperties!.push(
							cssProperty as CssCustomProperty
						);
					}
				}
			}
		},
	};
};
