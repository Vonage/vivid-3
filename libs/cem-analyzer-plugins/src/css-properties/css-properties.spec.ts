import * as path from 'path';
import * as fs from 'fs';
import * as glob from 'glob';
import * as ts from '@custom-elements-manifest/analyzer/node_modules/typescript';
import { create } from '@custom-elements-manifest/analyzer/src/create.js';
import { cssPropertiesPlugin } from './css-properties';

const fixturesPath = path.join(__dirname, '__fixtures__');
const testCases = fs.readdirSync(fixturesPath);

test.each(testCases)(`Testcase %s`, async (testCase) => {
	const moduleFiles = glob.sync(
		path.join(fixturesPath, testCase, '/**/*.{ts,js}')
	);
	const modules = moduleFiles.map((moduleFile) =>
		ts.createSourceFile(
			path.basename(moduleFile),
			fs.readFileSync(moduleFile, 'utf-8'),
			ts.ScriptTarget.ES2015,
			true
		)
	);
	process.chdir(path.join(fixturesPath, testCase));

	expect(
		create({ modules, plugins: [cssPropertiesPlugin()] })
	).toMatchSnapshot();
});
