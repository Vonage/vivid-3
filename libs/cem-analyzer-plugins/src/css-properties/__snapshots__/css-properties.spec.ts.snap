// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testcase default 1`] = `
{
  "modules": [
    {
      "declarations": [
        {
          "cssProperties": [
            {
              "name": "--my-prop",
            },
            {
              "description": "My Prop",
              "name": "--my-prop-desc",
            },
            {
              "default": "red",
              "name": "--my-prop-def",
            },
            {
              "default": "red",
              "description": "My Prop",
              "name": "--my-prop-def-desc",
            },
            {
              "name": "--my-prop-alt",
            },
          ],
          "customElement": true,
          "description": "",
          "kind": "class",
          "name": "MyElement",
          "superclass": {
            "name": "HTMLElement",
          },
        },
      ],
      "exports": [
        {
          "declaration": {
            "module": "my-element.ts",
            "name": "MyElement",
          },
          "kind": "js",
          "name": "MyElement",
        },
      ],
      "kind": "javascript-module",
      "path": "my-element.ts",
    },
    {
      "declarations": [
        {
          "customElement": true,
          "description": "",
          "kind": "class",
          "name": "OtherElement",
          "superclass": {
            "name": "HTMLElement",
          },
        },
      ],
      "exports": [
        {
          "declaration": {
            "module": "other-element.ts",
            "name": "OtherElement",
          },
          "kind": "js",
          "name": "OtherElement",
        },
      ],
      "kind": "javascript-module",
      "path": "other-element.ts",
    },
  ],
  "readme": "",
  "schemaVersion": "1.0.0",
}
`;
